core/workspace_drag_surface_svg.js

```diff
@@ -0,0 +1,181 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview An SVG that floats on top of the workspace.
+ * Blocks are moved into this SVG during a drag, improving performance.
+ * The entire SVG is translated, so the blocks are never repainted during drag.
+ * @author katelyn@google.com (Katelyn Mann)
+ */
+
+'use strict';
+
+goog.provide('Blockly.WorkspaceDragSurfaceSvg');
+
+goog.require('Blockly.utils');
+
+goog.require('goog.asserts');
+goog.require('goog.math.Coordinate');
+
+/**
+ * Class for a Drag Surface SVG.
+ * @param {Element} container Containing element.
+ * @constructor
+ */
+Blockly.workspaceDragSurfaceSvg = function(container) {
+  this.container_ = container;
+};
+
+/**
+ * The SVG drag surface. Set once by Blockly.workspaceDragSurfaceSvg.createDom.
+ * @type {Element}
+ * @private
+ */
+Blockly.workspaceDragSurfaceSvg.prototype.SVG_ = null;
+
+/**
+ * SVG group inside the drag surface. This is where blocks are moved to.
+ * @type {Element}
+ * @private
+ */
+Blockly.workspaceDragSurfaceSvg.prototype.dragGroup_ = null;
+
+/**
+ * Containing HTML element; parent of the workspace and the drag surface.
+ * @type {Element}
+ * @private
+ */
+Blockly.workspaceDragSurfaceSvg.prototype.container_ = null;
+
+
+/**
+ * Create the drag surface and inject it into the container.
+ */
+Blockly.workspaceDragSurfaceSvg.prototype.createDom = function() {
+  if (this.SVG_) {
+    return;  // Already created.
+  }
+
+  /**
+  * FILL dom structure in!
+  */
+  this.SVG_ = Blockly.utils.createSvgElement('svg', {
+    'xmlns': Blockly.SVG_NS,
+    'xmlns:html': Blockly.HTML_NS,
+    'xmlns:xlink': 'http://www.w3.org/1999/xlink',
+    'version': '1.1',
+    'class': 'blocklyWsDragSurface'
+  }, null);
+  this.container_.appendChild(this.SVG_);
+};
+
+
+/**
+ * Translate the entire drag surface during a drag.
+ * We translate the drag surface instead of the blocks inside the surface
+ * so that the browser avoids repainting the SVG.
+ * Because of this, the drag coordinates must be adjusted by scale.
+ * @param {Number} x X translation for the entire surface
+ * @param {Number} y Y translation for the entire surface
+ * @package
+ */
+Blockly.workspaceDragSurfaceSvg.prototype.translateSurface = function(x, y) {
+  // Force values to have two decimal points.
+  // This is a work-around to prevent a bug in Safari, where numbers close to 0
+  // are sometimes reported as something like "2.9842794901924208e-12".
+  // That is incompatible with translate3d, causing bugs.
+  x = x.toFixed(0);
```

Comment:
Comment doesn't match code: The comment says two decimal points, but the code says no decimal points.